docs/proposals/cascading-deletion.md

```diff
@@ -0,0 +1,364 @@
+8. The "orphan" finalizer does *not* take any action, because the observed deletion shows `D1` has an empty Finalizers map.
+9. The Propagator of the GC observes the deletion of `D1`. It deletes `D1` from the DAG. It adds its children, replicaset `R1`, to the *dirty queue*.
+10. The Garbage Processor of the GC dequeues `R1` from the *dirty queue*. It finds `R1` has a parent reference pointing to `D1`, and `D1` no longer exists, so it requests API server to delete `R1`, with `DeleteOptions.OrphanChildren=false`. (The Garbage Processor should always set this field to false.)
+11. The API server updates `R1` to remove the "orphan" finalizer if it's in the `R1`'s Finalizers map. Then the API server deletes `R1`, as `R1` has an empty Finalizers map.
+12. The Propagator of the GC observes the deletion of `R1`. It deletes `R1` from the DAG. It adds its children, Pods `P1`~`Pn`, to the *Dirty Queue*.
+13. The Garbage Processor of the GC dequeues `Px` (1 <= x <= n) from the *Dirty Queue*. It finds that `Px` have a parent reference pointing to `D1`, and `D1` no longer exists, so it requests API server to delete `Px`, with `DeleteOptions.OrphanChildren=false`.
+14. API server deletes the Pods.
+
+  In case the user wants to orphan `D1`'s descendants, then
+
+7. The user deletes the deployment `D1`, with `DeleteOptions.OrphanChildren=true`.
+8. The API server first updates `D1`, with DeletionTimestamp=now and DeletionGracePeriodSeconds=0, increments the Generation by 1, and add the "orphan" finalizer to ObjectMeta.Finalizers if it's not present yet. The API server does not delete `D1`, because its Finalizers map is not empty.
+9. The deployment controller observes the update, and acknowledges by updating the `D1`'s ObservedGeneration. The deployment controller won't create more replicasets on `D1`'s behalf.
+10. The "orphan" finalizer observes the update, and notes down the Generation. It waits until the ObservedGeneration becomes equal to or greater than the noted Generation. Then it updates `R1` to remove `D1` from its ParentReferences. At last, it updates `D1`, removing itself from `D1`'s Finalizers map.
+11. The API server handles the update of `D1`, because *i)* DeletionTimestamp is non-nil, *ii)*  the DeletionGracePeriodSeconds is zero, and *iii)* the last finalizer is removed from the Finalizers map, API server deletes `D1`.
+12. The Propagator of the GC observes the deletion of `D1`. It deletes `D1` from the DAG. It adds its child, replicaset `R1`, to the *Dirty Queue*.
+13. The Garbage Processor of the GC dequeues `R1` from the *Dirty Queue* and skips it, because its ParentReferences is empty.
+
+## Life of a ConfigMap
+
+1. User creates a ConfigMap `C1`.
+2. User creates a Deployment `D1`, which refers `C1` in the pod template.
+3. The deployment controller has observed the creation of `D1`. It creates replicaset `R1`, which also refer `C1` in the pod template. It then updates `C1`, adding `R1` to the ParentReferences.
```

Comment:
I consider GC of ConfigMaps for Deployment to be a separate feature. We should be able to use the general GC mechanism to do the actual GC, but setting up the correct parent references will be up to Deployment. Do you think you'll be able to tackle that as well?

I do not want to add ConfigMap and/or Secret templates to Deployment. I want them to stay separate objects. 

I expect them to be mostly generated by tooling/automation, such as kubectl create configmap..., rather than hand-edited YAML. 

How to best convey to Deployment which ConfigMaps to parent under managed ReplicaSets is TBD. One possibility would be to parent all of them, but then they might be unexpectedly deleted if the user didn't specify other parents. Another could be to parent ones whose labels match the Deployment's selector, but that might be too subtle, and perhaps incompatible with selector generation. We could add another selector, but that seems more complex than I would like.

cc @janetkuo @pwittrock